usr.sbin/ypserv/ypserv/ypserv.8
.\"	$NetBSD: ypserv.8,v 1.19 2008/05/16 16:41:42 chuck Exp $
.\"
.\" Copyright (c) 1994 Mats O Jansson <moj@stacken.kth.se>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Mats O Jansson
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 20, 2006
.Dt YPSERV 8
.Os
.Sh NAME
.Nm ypserv
.Nd NIS server daemon
.Sh SYNOPSIS
.Nm
.Op Fl dfl
.Op Fl p Ar port
.Sh DESCRIPTION
.Nm
is a fundamental part of the network information system called
.Tn NIS .
This server provides information from
.Tn NIS
maps to the
.Tn NIS
clients on the network.
.Pp
A
.Tn NIS
map is stored on the server as a
.Xr db 3
database.
A number of
.Tn NIS
maps is grouped together in a domain.
.Nm
determines the domains it serves by looking for a directory with
the domain name in
.Ar /var/yp .
.Pp
In an effort to improve the security of
.Tn NIS
(which has, historically, not been very good), this
.Nm
has support for libwrap-based access control.
See
.Xr hosts_access 5
for more information.
The
.Em daemon
used for access control is the name which
.Nm
was invoked as (typically
.Dq ypserv
).  If a host is not allowed to query this
.Tn NIS
server,
.Nm
will return the
.Tn NIS
result code YP_NODOM.
To avoid problems with DNS lookups causing
.Nm
to hang,
.Nm
disables DNS lookups for its client
.Xr hosts_access 5
lists.  The result is that
.Nm
can only use address based patterns.  This also means that wildcard patterns
such as LOCAL or KNOWN will not work.
.Pp
The process pid of the
.Nm
process can be found in the file
.Pa /var/run/ypserv.pid .
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl d
Use internet Domain Name System. If a query to map
.Pa hosts.byname
or
.Pa hosts.byaddr
fails, make a DNS query and return the result if successful.
.It Fl f
Run in the foreground.
.It Fl l
Enable logging of all requests.
.It Fl p Ar port
Bind to the specified
.Ar port
instead of dynamically allocating one.
.El
.Pp
All messages are sent to the system log with the facility LOG_DAEMON.
Error messages have the priority LOG_ERR.  Refused requests are logged
with the priority LOG_WARNING.  All other messages are logged with the
priority LOG_INFO.
.Sh FILES
.Bl -tag -width /var/run/ypserv.pid -compact
.It Pa /var/run/ypserv.pid
.El
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr hosts_access 5 ,
.Xr nis 8 ,
.Xr syslogd 8 ,
.Xr ypbind 8 ,
.Xr ypinit 8
.Sh AUTHORS
This implementation of
.Nm
was originally written by
.An Mats O Jansson
.Aq moj@stacken.kth.se .
The access control code was later re-written from scratch by
.An Jason R. Thorpe
.Aq thorpej@NetBSD.org .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1142.2.10 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: Add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: Add support for four more NuBus video cards, including the
		VillageTronic MacPicasso 340. [hauke 20061206]
	puc(4): Add support for Perle Systems 4 and 8 ports modem boards.
		[manu 20071209]
	arm: Add support for NPE Ethernet on IXP425. [scw 20061210]
	arm: Add support for IXP425 watchdog timer. [scw 20061210]
	libutil: Add flags_to_string(3) and string_to_flags(3),
		moved from ls(1). [christos 20061214]
	arm: Add support for i2c, i2s, pcmcia, and ohci of PXA2x0.
		Ported from OpenBSD as part of zaurus port by NONAKA Kimihiro.
		[peter 20061217]
	zaurus: Add support for Sharp C3x00 PDA. [ober 20061217]
	kernel: Merge yamt-splraseipl branch, which provides splraiseipl(9)
		and makeiplcookie(9). [yamt 20061221]
	sgimips: Add support for Set Engineering's GIO Fast Ethernet board.
		[rumble 20061223]
	sgimips: Add support for the SGI Light/Entry/Starter LG1/LG2
		framebuffers found in Indigo systems. [rumble 20061226]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	kernel, srtconfig(1): First import of the source-address-based routing
		pseudo-device.  [mouse 20061229]
	sgimips: Add i8254 timecounter(9) support for use on IP12.
		[rumble 20061229]
	mount_psshfs(8): Add puffs sshfs. [pooka 20061229]
	njata(4): Add support for Workbit CF32A CF adapter,
		aka BUFFALO RCF-CBA2.  [ichiro 20061231]
	amd64, i386: Report and work around some of the known errata for
		Athlon 64 and Opteron processors. [ad 20070101]
	i386: Add fatboot, sub-512 byte bootloader that can be written into
		the first sector of a FAT16 filesystem and will load
		the secondary bootloader from it. [dsl 20070101]
	kernel: Make mount(2) and unmount(2) use kauth(9) for security
		policy.  One incompatible change is that an unprivileged
		mount without the MNT_NOSUID and MNT_NODEV flags
		will now fail, whereas previously those flags were
		automatically added.  [elad 20070102]
	udf: Implement read-only access to UDF 2.50 and UDF 2.60 discs
		that use metadata partition mappings. [reinoud 20070104]
	i386: Add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface
		address changes is now supported - no more ntpd
		restarts when interface addresses are re-configured
		(WLAN/DSL/ppp/...). [kardel 20070106]
	config(1): Introduce a new -L option, generates a lint configuration.
		[cube 20070108]
	umass(4): Add a quirk for the Sony Cybershot USB camera.
		From Mihai Chelaru. [christos 20070109]
	kernel: Global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	gcc(1): Updated to version 4.1.2-20070110. [mrg 20070111]
	zaurus: Add MI todr(9) and timecounter(9) support. [ober 20070112]
	acpitools: Added from FreeBSD via fvdl. [christos 20070114]
	kernel: Add syscalls for POSIX Real-time library. [ad 20070116]
	adb(4): Add new and simplified MI ADB drivers. [macallan 20070117]
	macppc: Add a driver for the SGS 7433 mixer found in some G3 Macs.
		[macallan 20070117]
	viaenv(4): Add support for the VIA VT8231 hardware monitor and
		remove viapm(4) code. From OpenBSD. [xtraeme 20070120]
	scsictl(8): Added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	msk(4): Pull support for some devices and many bug fixes from OpenBSD.
		[msaitoh 20070130]
	uplcom(4): Add support for Willcom WS002IN PHS device
		(Prolific Technology PL2303X). [msaitoh 20070131]
	zaurus: Updates to allow Zaurus screen to rotate 90 degrees to a usable
		state with the keyboard. From OpenBSD. [ober,peter 20070202]
	security(8): Add support for per-user /tmp. [elad 20070204]
	ugensa(4): Add support Novatel Merlin, used in VZ V620.
		From John Nielsen on freebsd-mobile. [gdt 20070205]
	i386, amd64: Add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	find(1): Add -delete option from FreeBSD. [elad 20070206]
	rum(4): Add support for MELCO WLI-U2-SG54HP and PLANEX GW-US54Mini2.
		[elad 20070206]
	kernel: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	ubsa(4): Add support for CDMA modems sold by Eurotel (now O2).
		[salo 20070210]
	refuse: Added FUSE (file system in userspace) compatibility
		functionality, layered on top of libpuffs. [agc 20070211]
	i386: Added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	shark: Add generic softintr(9) support. [matt 20070218]
	kernel: Add code to support per-system call statistics. [dsl 20070218]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	sgimips: Support the sq(4) interface on the Challenge S's IOPLUS

usr.sbin/ypserv/ypserv/ypserv.c
/*	$NetBSD: ypserv.c,v 1.22 2008/05/16 16:41:42 chuck Exp $	*/

/*
 * Copyright (c) 1994 Mats O Jansson <moj@stacken.kth.se>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Mats O Jansson
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__RCSID("$NetBSD: ypserv.c,v 1.22 2008/05/16 16:41:42 chuck Exp $");
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>

#include <err.h>
#include <netdb.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <util.h>
#include <stdarg.h>
#include <errno.h>

#include <rpc/rpc.h>
#include <rpc/xdr.h>
#include <rpc/pmap_clnt.h>

#include <rpcsvc/yp_prot.h>

#include "ypdef.h"
#include "ypserv.h"

#ifdef LIBWRAP
#include <tcpd.h>

int allow_severity = LOG_DAEMON | LOG_INFO;
int deny_severity = LOG_DAEMON | LOG_WARNING;

/* XXX For ypserv_proc.c -- NOT THREAD SAFE!  (like any of this code is) */
const char *clientstr;
const char *svcname;
#endif /* LIBWRAP */

int	usedns;
#ifdef DEBUG
static int	foreground = 1;
#else
static int	foreground;
#endif

#ifdef LIBWRAP
int	lflag;
#endif

static struct bindsock {
	sa_family_t family;
	int type;
	int proto;
	const char *name;
} socklist[] = {
	{ AF_INET, SOCK_DGRAM, IPPROTO_UDP, "udp" },
	{ AF_INET, SOCK_STREAM, IPPROTO_TCP, "tcp" },
	{ AF_INET6, SOCK_DGRAM, IPPROTO_UDP, "udp6" },
	{ AF_INET6, SOCK_STREAM, IPPROTO_TCP, "tcp6" },
};

static void	usage(void) __dead;
static int	bind_resv_port(int, sa_family_t, in_port_t);
void		ypserv_sock_hostname(struct host_info *host);

static void
_msgout(int level, const char *msg, ...)
{
	va_list ap;
	va_start(ap, msg);
	if (foreground)
                vwarnx(msg, ap);
        else
		vsyslog(level, msg, ap);
	va_end(ap);
}

void ypserv_sock_hostname(struct host_info *host)
{
	host->name[0] = 0;
}

static void
ypprog_2(struct svc_req *rqstp, SVCXPRT *transp)
{
	union {
		char * ypproc_domain_2_arg;
		char * ypproc_domain_nonack_2_arg;
		struct ypreq_key ypproc_match_2_arg;
		struct ypreq_nokey ypproc_first_2_arg;
		struct ypreq_key ypproc_next_2_arg;
		struct ypreq_xfr ypproc_xfr_2_arg;
		struct ypreq_nokey ypproc_all_2_arg;
		struct ypreq_nokey ypproc_master_2_arg;
		struct ypreq_nokey ypproc_order_2_arg;
		char * ypproc_maplist_2_arg;
	} argument;
	void *argp = &argument;
	char *result;
	xdrproc_t xdr_argument, xdr_result;
	void *(*local)(void *, struct svc_req *);
#ifdef LIBWRAP
	struct request_info req;
	struct sockaddr *caller;
#define	SVCNAME(x)	svcname = x
#else
#define	SVCNAME(x)	/* nothing */
#endif

#ifdef LIBWRAP
	caller = svc_getrpccaller(transp)->buf;
	(void)request_init(&req, RQ_DAEMON, getprogname(), RQ_CLIENT_SIN,
	    caller, NULL);
	sock_methods(&req);

	/*
	 * Do not do hostname lookups!  This avoids possible delays due
	 * to DNS, preventing a possible DoS attack, as well as possible
	 * circular lookups (e.g. a hostname lookup requiring a request
	 * to ourselves).
	 */
	req.hostname = ypserv_sock_hostname;
#endif

	switch (rqstp->rq_proc) {
	case YPPROC_NULL:
		xdr_argument = xdr_void;
		xdr_result = xdr_void;
		local = ypproc_null_2_svc;
		SVCNAME("null_2");
		break;

	case YPPROC_DOMAIN:
		xdr_argument = xdr_ypdomain_wrap_string;
		xdr_result = xdr_bool;
		local = ypproc_domain_2_svc;
		SVCNAME("domain_2");
		break;

	case YPPROC_DOMAIN_NONACK:
		xdr_argument = xdr_ypdomain_wrap_string;
		xdr_result = xdr_bool;
		local = ypproc_domain_nonack_2_svc;
		SVCNAME("domain_nonack_2");
		break;

	case YPPROC_MATCH:
		xdr_argument = xdr_ypreq_key;
		xdr_result = xdr_ypresp_val;
		local = ypproc_match_2_svc;
		SVCNAME("match_2");
		break;

	case YPPROC_FIRST:
		xdr_argument = xdr_ypreq_nokey;
		xdr_result = xdr_ypresp_key_val;
		local = ypproc_first_2_svc;
		SVCNAME("first_2");
		break;

	case YPPROC_NEXT:
		xdr_argument = xdr_ypreq_key;
		xdr_result = xdr_ypresp_key_val;
		local = ypproc_next_2_svc;
		SVCNAME("next_2");
		break;

	case YPPROC_XFR:
		xdr_argument = xdr_ypreq_xfr;
		xdr_result = xdr_ypresp_xfr;
		local = ypproc_xfr_2_svc;
		SVCNAME("xfer_2");
		break;

	case YPPROC_CLEAR:
		xdr_argument = xdr_void;
		xdr_result = xdr_void;
		local = ypproc_clear_2_svc;
		SVCNAME("clear_2");
		break;

	case YPPROC_ALL:
		xdr_argument = xdr_ypreq_nokey;
		xdr_result = xdr_ypresp_all;
		local = ypproc_all_2_svc;
		SVCNAME("all_2");
		break;

	case YPPROC_MASTER:
		xdr_argument = xdr_ypreq_nokey;
		xdr_result = xdr_ypresp_master;
		local = ypproc_master_2_svc;
		SVCNAME("master_2");
		break;

	case YPPROC_ORDER:
		xdr_argument = xdr_ypreq_nokey;
		xdr_result = xdr_ypresp_order;
		local = ypproc_order_2_svc;
		SVCNAME("order_2");
		break;

	case YPPROC_MAPLIST:
		xdr_argument = xdr_ypdomain_wrap_string;
		xdr_result = xdr_ypresp_maplist;
		local = ypproc_maplist_2_svc;
		SVCNAME("maplist_2");
		break;

	default:
		svcerr_noproc(transp);
		return;
	}

#ifdef LIBWRAP
	clientstr = eval_client(&req);

	if (hosts_access(&req) == 0) {
		syslog(deny_severity,
		    "%s: refused request from %.500s", svcname, clientstr);
		svcerr_auth(transp, AUTH_FAILED);
		return;
	}
#endif

	(void)memset(&argument, 0, sizeof (argument));
	if (!svc_getargs(transp, xdr_argument, argp)) {
		svcerr_decode(transp);
		return;
	}
	result = (*local)(&argument, rqstp);
	if (result != NULL && !svc_sendreply(transp, xdr_result, result)) {
		svcerr_systemerr(transp);
	}
	if (!svc_freeargs(transp, xdr_argument, argp)) {
		_msgout(LOG_ERR, "unable to free arguments");
		exit(1);
	}
	return;
}

/*
 * limited NIS version 1 support: the null, domain, and domain_nonack
 * request/reply format is identical between v1 and v2.  SunOS4's ypbind
 * makes v1 domain_nonack calls.
 */
static void
ypprog_1(struct svc_req *rqstp, SVCXPRT *transp)
{
	switch (rqstp->rq_proc) {
	case YPPROC_NULL:
	case YPPROC_DOMAIN:
	case YPPROC_DOMAIN_NONACK:
		ypprog_2(rqstp, transp);
		return;

	default:
		svcerr_noproc(transp);
		return;
	}
}

int
main(int argc, char *argv[])
{
	SVCXPRT *xprt;
	struct netconfig *cfg = NULL;
	int s;
	struct sigaction sa;
	struct bindsock *bs;
	in_port_t port = 0;
	int ch, xcreated = 0, one = 1;

	setprogname(argv[0]);

#ifdef LIBWRAP
#define	GETOPTSTR	"dflp:"
#else
#define	GETOPTSTR	"dfp:"
#endif
	while ((ch = getopt(argc, argv, GETOPTSTR)) != -1) {
		switch (ch) {
		case 'd':
			usedns = 1;
			break;
		case 'f':
			foreground = 1;
			break;
		case 'p':
			port = atoi(optarg);
			break;
#ifdef LIBWRAP
		case 'l':
			lflag = 1;
			break;
#endif
		default:
			usage();
		}
	}

#undef GETOPTSTR

	/* This program must be run by root. */
	if (geteuid() != 0)
		errx(1, "must run as root");

	if (foreground == 0 && daemon(0, 0))
		err(1, "can't detach");

	openlog("ypserv", LOG_PID, LOG_DAEMON);
	syslog(LOG_INFO, "starting");
	(void)pidfile(NULL);

	(void) rpcb_unset((u_int)YPPROG, (u_int)YPVERS, NULL);
	(void) rpcb_unset((u_int)YPPROG, (u_int)YPVERS_ORIG, NULL);


	ypdb_init();	/* init db stuff */

	sa.sa_handler = SIG_IGN;
	sa.sa_flags = SA_NOCLDWAIT;
	if (sigemptyset(&sa.sa_mask)) {
		_msgout(LOG_ERR, "sigemptyset: %s", strerror(errno));
		exit(1);
	}
	if (sigaction(SIGCHLD, &sa, NULL)) {
		_msgout(LOG_ERR, "sigaction: %s", strerror(errno));
		exit(1);
	}

	for (bs = socklist;
	    bs < &socklist[sizeof(socklist) / sizeof(socklist[0])]; bs++) {

		if ((s = socket(bs->family, bs->type, bs->proto)) == -1)
			continue;

		if (bs->family == AF_INET6) {
			/*
			 * We're doing host-based access checks here, so don't
			 * allow v4-in-v6 to confuse things.
			 */
			if (setsockopt(s, IPPROTO_IPV6,
			    IPV6_V6ONLY, &one, sizeof(one)) == -1) {
				_msgout(LOG_ERR,
				    "can't disable v4-in-v6 on %s socket",
				    bs->name);
				exit(1);
			}
		}

		if ((cfg = getnetconfigent(bs->name)) == NULL) {
			_msgout(LOG_ERR,
			    "unable to get network configuration for %s port",
			    bs->name);
			goto out;
		}

		if (bind_resv_port(s, bs->family, port) != 0)
			goto out;

		if (bs->type == SOCK_STREAM) {
			(void)listen(s, SOMAXCONN);
			xprt = svc_vc_create(s, 0, 0);
		} else {
			xprt = svc_dg_create(s, 0, 0);
		}

		if (xprt == NULL) {
			_msgout(LOG_WARNING, "unable to create %s service",
			    bs->name);
			goto out;
		}
		if (svc_reg(xprt, (u_int)YPPROG, (u_int)YPVERS_ORIG, ypprog_1,
		    cfg) == 0 ||
		    svc_reg(xprt, (u_int)YPPROG, (u_int)YPVERS, ypprog_2,
		    cfg) == 0) {
			_msgout(LOG_WARNING, "unable to register %s service",
			    bs->name);
			goto out;
		}
		xcreated++;
		freenetconfigent(cfg);
		continue;
out:
		if (s != -1)
			(void)close(s);
		if (cfg) {
			freenetconfigent(cfg);
			cfg = NULL;
		}
	}

	if (xcreated == 0) {
		_msgout(LOG_ERR, "unable to create any services");
		exit(1);
	}

	svc_run();
	_msgout(LOG_ERR, "svc_run returned");
	exit(1);
	/* NOTREACHED */
}

static void
usage(void)
{

#ifdef LIBWRAP
#define	USAGESTR	"Usage: %s [-d] [-l] [-p <port>]\n"
#else
#define	USAGESTR	"Usage: %s [-d] [-p <port>]\n"
#endif

	(void)fprintf(stderr, USAGESTR, getprogname());
	exit(1);

#undef USAGESTR
}

/*
 * _yp_invalid_map: check if given map name isn't legal.
 * returns non-zero if invalid
 *
 * XXX: this probably should be in libc/yp/yplib.c
 */
int
_yp_invalid_map(const char *map)
{
	if (map == NULL || *map == '\0')
		return 1;

	if (strlen(map) > YPMAXMAP)
		return 1;

	if (strchr(map, '/') != NULL)
		return 1;

	return 0;
}

static int
bind_resv_port(int sock, sa_family_t family, in_port_t port)
{
	struct sockaddr *sa;
	struct sockaddr_in sasin;
	struct sockaddr_in6 sasin6;

	switch (family) {
	case AF_INET:
		(void)memset(&sasin, 0, sizeof(sasin));
		sasin.sin_len = sizeof(sasin);
		sasin.sin_family = family;
		sasin.sin_port = htons(port);
		sa = (struct sockaddr *)(void *)&sasin;
		break;
	case AF_INET6:
		(void)memset(&sasin6, 0, sizeof(sasin6));
		sasin6.sin6_len = sizeof(sasin6);
		sasin6.sin6_family = family;
		sasin6.sin6_port = htons(port);
		sa = (struct sockaddr *)(void *)&sasin6;
		break;
	default:
		_msgout(LOG_ERR, "Unsupported address family %d", family);
		return -1;
	}
	if (bindresvport_sa(sock, sa) == -1) {
		_msgout(LOG_ERR, "Cannot bind to reserved port %d (%s)", port,
		    strerror(errno));
		return -1;
	}
	return 0;
}
